src/wavm/wavm.cc

```diff
@@ -90,16 +91,35 @@ namespace {
       WAVM::Runtime::catchRuntimeExceptions(                                                       \
           [&] { _x; },                                                                             \
           [&](WAVM::Runtime::Exception *exception) {                                               \
-            auto description = describeException(exception);                                       \
-            _wavm->fail(FailState::RuntimeError,                                                   \
-                        "Function: " + std::string(function_name) + " failed: " + description);    \
-            destroyException(exception);                                                           \
+            _wavm->fail(FailState::RuntimeError, getFailMessage(function_name, exception));        \
             throw std::exception();                                                                \
           });                                                                                      \
     } catch (...) {                                                                                \
     }                                                                                              \
   } while (0)
 
+std::string getFailMessage(std::string_view function_name, WAVM::Runtime::Exception *exception) {
+  std::string message = "Function " + std::string(function_name) +
+                        " failed: " + WAVM::Runtime::describeExceptionType(exception->type) +
+                        "\nProxy-Wasm plugin in-VM backtrace:\n";
+  std::vector<std::string> callstack_descriptions =
+      WAVM::Runtime::describeCallStack(exception->callStack);
+
+  // Since the first frame is on host and useless for developers, e.g.: `host!envoy+112901013`
+  // we start with index 1 here
+  for (size_t i = 1; i < callstack_descriptions.size(); i++) {
+    std::string description = callstack_descriptions[i];
+    if (description.find("wasm!") == std::string::npos) {
+      // end of WASM's call stack
+      break;
+    }
+    message += "  " + std::to_string(i) + ": " + description + "\n";
```

Comment:
Actually, one more thing. Could you add leading spaces here, so that single-digit and double-digit numbers have the same width? Right now, it's going to look a bit ugly once we hit 10:
```
[2020-10-28 14:24:10.551][979699][trace][wasm] [source/extensions/common/wasm/wasm_vm.cc:29] Function proxy_on_tick failed: wavm.reachedUnreachable
Proxy-Wasm plugin in-VM backtrace:
  1: wasm!helloworld!runtime._panic+7
  2: wasm!helloworld!command-line-arguments.C+2
  3: wasm!helloworld!command-line-arguments.B+0
  4: wasm!helloworld!command-line-arguments.A+0
  5: wasm!helloworld!(*command-line-arguments.helloWorld).OnTick+74
  6: wasm!helloworld!proxy_on_tick+0
  7: wasm!helloworld!proxy_on_tick+0
  8: wasm!helloworld!proxy_on_tick+0
  9: wasm!helloworld!proxy_on_tick+0
  10: wasm!helloworld!proxy_on_tick+0
```

I think you could use `std::ostringstream` with `<< std::setfill(' ') << std::setw(3)`, I'm not sure if there is any easier way to do that in C++.

Reply:
yeah, that's true. I'm on it

Reply:
done

![Screenshot from 2020-10-29 19-48-34](https://user-images.githubusercontent.com/13513977/97558674-c79a0400-1a1f-11eb-907e-0b4cb8a9dd83.png)